### DISCUSSION TITLE
Splitting up / refactoring axes classes

### DISCUSSION THREAD

**Discussion by bryevdv** [bokeh/bokeh#14042] [Internals and design (Q&A)]
Axis and its subclasses are currently some of the gnarliest corners of BokehJS. There are obvious problems like the need to use `any` in categorical axes because currenty `Coords` are assume to be numeric everywhere else. Mostly it's just overly complicated and duplicative of other parts of the codebase. I think that mostly stems from `Axis` having to do too much in the early days, since there were not other pieces in place. 

I think if I had to do things again, Axis would look more like this:
```python
axis.minor_ticks.line_color  = "red"       # "rug" annotation glyph

axis.major_ticks.tick_in = 3               # "rug" annotation glyph

axis.major_labels.text_font_size = "12pt"  # text label glyph

axis.rule.line_width = 2                   # span annotation glyph
```

But I guess a change like that would be Bokeh 4.0 territory. What can we do in the mean time to try and tease things apart? Some ideas:

* split out lightweight internal classes that are responsible for reacting to subsets of the existing "large" configuration set of `Axis` 
* actually use glyphs internally for rendering

@mattpap do you already have any plans/roadmap around re-using / de-duplicating code, especially wrt to Axes?

**Comment by mattpap** [comment]
> `axis.minor_ticks.line_color`

On one hand I always have wished these properties were organized like this. However, then it wouldn't be possible to initialize axis models with a single call, like currently `LinearAxis(minor_tick_line_color="red")`. Instead we would have to use something like `LinearAxis(minor_ticks=LineVisuals(color="red"))` or even worse a `dict()`.

> actually use glyphs internally for rendering

Glyphs have well defined meaning for vectorized data and ill defined semantics for scalars. They come will a lot of complexity and associated cost related to vectorization and other glyph specific aspects (data inheritance, spatial indexing, glyph renderers vs glyphs, etc.). To avoid all of this, there's a new API for creating basic shapes in PR #13279.

> `axis.rule.line_width = 2                   # span annotation glyph`

In some form, this has been on my TODO list as work adjacent to my previous SDG work on glyph decorations, because I wanted to add arrowheads to axis rules.

> Do you already have any plans/roadmap around re-using / de-duplicating code, especially wrt to Axes?

Axes will change quite significantly with the ongoing and upcoming CZI R6 work, but I don't have a lot to say about this right now. I don't have any non-specific refactorings planned, because I don't find refactorings particularly useful, that don't emerge from a well defined set of requirements.
